### DISCUSSION TITLE
why the change from virtualItem.measureElement -> virtualizer.measureElement + data-index attr?

### DISCUSSION THREAD

**Discussion by kamikazePT** [TanStack/virtual#617] [Q&A] [answered]
I was using beta.2 in a production environment for quite a while now, and trying to upgrade locally broke a dynamic sized list of cards (similar to a news feed).
After digging into it, I've noticed that `measureElement` function was removed from the virtualItem, and now we have to call that function from virtualizer instance, appending an attribute data-index to the DOM node.

What was the motivation for this change? to keep VirtualItem with only state and all the logic inside Virtualizer?
I think the previous approach found in beta.2 made for a cleaner API, since you would just loop VirtualItems and simply append a ref to the underlying element. Now we need to keep in mind that besides the ref, we also need an extra attribute

**Comment by piecyk** [comment, marked as answer]
> What was the motivation for this change?

Yep, that was the case, reading the size of element when scrolling is expensive, and with not stale refs we would have extra calls 

> If the ref callback is defined as an inline function, it will get called twice during updates, first with null and then again with the DOM element. This is because a new instance of the function is created with each render, so React needs to clear the old ref and set up the new one. You can avoid this by defining the ref callback as a bound method on the class, but note that it shouldn’t matter in most cases.

 https://reactjs-tw.github.io/docs/refs-and-the-dom.html#caveats-with-callback-refs

Plus composing two measureElement like in grid example, row and column with item measureElement is not possible to keep it stale.

**Reply by kamikazePT** [reply]
thanks 👍
